code/controllers/subsystem/machinery.dm

```diff
@@ -26,16 +26,24 @@ var/list/machines = list()
 
 // This is to allow the near identical fast machinery process to use it.
 /datum/subsystem/machinery/proc/get_currenrun()
-	return machines.Copy()
+	currentrun_index = machines.len
+	currentrun.len = currentrun_index
+	if (currentrun_index)
+		for(var/c in 1 to currentrun_index)
+			currentrun[c] = machines[c]
```

Comment:
There is absolutely no way this is faster than `Copy()`

Reply:
OK yeah I tested it and this is **vastly** slower than just using `Copy()`
Like 10+ times slower, and probably way more than that in this case

Reply:
I feel like if you're seeing an improvement from this, it must mean that it's not doing the correct thing.

Reply:
It seems like there is some variability in the effect but if I:
-compile it with these changes
-host it locally and join
-clear the profiler immediately after the engines subsystem pauses,
-wait till there are over ~170 calls on the subsystem/machinery/fire()
-divide total CPU by calls
it results in ~0.00690
and if I then stop the server, close dream seeker, close the game, switch to Bleeding-Edge, do the previous steps again,
it results in ~0.00853

I think maybe it being a long list is what makes the difference.

Reply:
It being a long list actually makes it *slower* compared to `Copy()`.
You need to test changes individually, not all at once. Testing all of your changes tells you nothing about any individual one.

Reply:
On top of Bleeding-Edge, if I change it to

```
/datum/subsystem/machinery/proc/get_currenrun()
	currentrun_index = machines.len
	return machines.Copy()

/datum/subsystem/machinery/fire(resumed = FALSE)
	if (!resumed)
		currentrun = get_currenrun()

	var/obj/machinery/M
	var/c = currentrun_index
	while (c)
		M = currentrun[c]
		c--

		if (!M || M.gcDestroyed || M.timestopped)
			continue

		if (M.process() == PROCESS_KILL)
			M.inMachineList = 0
			machines.Remove(M)
			continue

		if (M.use_power)
			M.auto_use_power()

		if (MC_TICK_CHECK)
			break

	currentrun_index = c
```
It comes out to 0.00678
Which is in the same range as the other things. So I guess the main thing that was using resources wasn't the list copy, but instead the iterative modifying of `currentrun.len`.

Reply:
Yes, it makes sense that iterating `currentrun` rather than consuming it would be faster.
Also, you can specify a range in `Copy()` if you want to retain the behavior of only copying up to `currentrun_index`.

Reply:
Seems like it takes (scaled numbers):
 `currentrun.len--` :
1.34 Self CPU/call

 `c--` 
0.868 Self CPU/call
So I'll keep `.Copy()` and instead just change it to that.

Reply:
Global vars are as fast as local vars. Datum vars are equally fast regardless of whether or not they're on `src`. Also, `static` datum vars are as fast as local vars, provided the variable is typecast. (This is the only situation in which typecasting a variable has any runtime effect lol) (It reads the static var from the type the variable is cast to, not the actual value of the variable)
Also, if you're wondering about moving the variables back out into `static` vars to avoid redeclaration: I was not able to measure a performance hit from redeclaration anyway. I suspect the variable is reused internally.

Reply:
> Datum vars are equally fast regardless of whether or not they're on src.

Oh I had been assuming that somehow being on src was faster.

So for example using the datum var "chem_temp" repeatedly in the thermal dissipation PR is actually slower than if I was to declare something else  = chem_temp and then use that, as long as you used it enough times that the speed boost would be outperformed by the extra declaration steps.
On that note.. I wonder how many times you have to use the local var instead of the datum var to make up for the declaration.

Reply:
Datum vars are slower to access than local vars because, despite pretending not to be, BYOND is actually a duck-typed language. So datum vars have to be looked up *by name* rather than just by a pointer internally.

> On that note.. I wonder how many times you have to use the local var instead of the datum var to make up for the declaration.

It's a bit harder to test the impact of declaration than *re*declaration, but my testing seems to show somewhere in the range of 3-5. It's possible that that's wrong or that it's more complicated than what I was testing, though.

Reply:
That clears a bunch of stuff up, thanks.